third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -307,16 +315,27 @@ impl Loader {
 
         let mut scripts = self.scripts.write();
         let main = match scripts.get(&hash_value) {
-            Some(cached) => cached,
+            Some(ScriptCacheEntry::Verified(script)) => script.entry_point(),
+            Some(ScriptCacheEntry::Deserialized(compiled_script)) => {
+                self.verify_script(compiled_script.as_ref(), data_store, module_store)?;
+                let script = Script::new(
+                    Arc::clone(compiled_script),
+                    &hash_value,
+                    module_store,
+                    &self.name_cache,
+                )?;
+                scripts.insert_verified_script(hash_value, script)
+            },
             None => {
-                let ver_script = self.deserialize_and_verify_script(
-                    script_blob,
-                    hash_value,
-                    data_store,
+                let compiled_script = self.deserialize_script(script_blob)?;
```

Comment:
First, would insert deserialized, then re-use the handling for verification (deserialized to verified). Makes me think that these could become the APIs of the script cache- e.g. if we never insert verified directly.